Revamp of the binding layer.

bosilca opened the pull request:
Rework the bindings. The main idea is to inherit the bindings from the batch scheduler, and then work from there.

bosilca committed:
Revamp of the binding (and the reporting) layer.

Follow the process binding provided by the batch scheduler or process manager.
If the application is started with mpirun, follow the bindings provided by the
mpirun command.

Add an MCA parameter (runtime_report_bindings) to report the final bindings for
each virtual process and thread. Report the binding in both logical and
physical notation.

Signed-off-by: George Bosilca <bosilca@icl.utk.edu>

devreal commented on parsec/parsec.c:
```diff
@@ -477,44 +480,53 @@ parsec_context_t* parsec_init( int nb_cores, int* pargc, char** pargv[] )
+    parsec_mca_param_reg_int_name("runtime", "num_cores", "The total number of cores to be used by the runtime (-1 for all available)",
+                                  false, false, parsec_runtime_max_number_of_cores, &parsec_runtime_max_number_of_cores);
+    parsec_mca_param_reg_int_name("runtime", "allow_pu",
+                                  "Allow threads to bind to PU (processing unit) instead of cores.",
```
What's a PU? A socket?

bosilca replied:
processing unit, a logical unit bound to a physical core. In general it is a core, except when there is hyper-threading when the PU is a hyper-thread and several hyperthreads make a core.

devreal commented on parsec/parsec.c:
```diff
@@ -2293,17 +2277,43 @@ void parsec_usage(void)
+    }
+    /* follow the allowed cpuset, but module the number of allowed compute resources */
+    idx = idx % hwloc_bitmap_weight(context->cpuset_allowed_mask);
+#if 0
```
Remove stale code?

devreal reviewed (commented):
Much needed :+1: two comments inline, otherwise LGTM

therault commented:
I'm trying it on leconte.

```
mpirun -np 2 --map-by socket hwloc-info --restrict binding package:0
Package L#0
 [...]
 cpuset = 0x0000ffff,0xf00000ff,0xfff00000
 complete cpuset = 0x0000ffff,0xf00000ff,0xfff00000
 allowed cpuset = 0x0000ffff,0xf00000ff,0xfff00000
 nodeset = 0x00000002
 complete nodeset = 0x00000002
 allowed nodeset = 0x00000002
 [...]
Package L#0
 [...]
 cpuset = 0x0fffff00,0x000fffff
 complete cpuset = 0x0fffff00,0x000fffff
 allowed cpuset = 0x0fffff00,0x000fffff
 nodeset = 0x00000001
 complete nodeset = 0x00000001
 allowed nodeset = 0x00000001
  [...]
```
Which I interpret as when running `mpirun -np 2 --map-by socket` on this machine (`module load openmpi`), there are two processes, and rank 0 should use a different set of cores than rank 1 (and the cpuset seems complicated, but it does seem exclusive).

Now, I run a parsec test with this PR:
```
mpirun -np 2 --map-by socket ./tests/apps/stencil/testing_stencil_1D -M 40960 -N 40960 -t 16 -T 16 -P 2
Process binding [rank 0]: cpuset [ALLOWED  ]: 0x0fffff00,0x000fffff
Process binding [rank 0]: cpuset [USED     ]: 0x000fffff
Process binding [rank 0]: cpuset [FREE     ]: 0x0fffff00,0x0
W@00000 parsec_hwloc: couldn't bind to mask cpuset  0x0
Process binding [rank 0]: cpuset [ALLOWED  ]: 0x0000ffff,0xf00000ff,0xfff00000
Process binding [rank 0]: cpuset [USED     ]: 0x000000ff,0xfff00000
Process binding [rank 0]: cpuset [FREE     ]: 0x0000ffff,0xf0000000,0x0
W@00001 parsec_hwloc: couldn't bind to mask cpuset  0x0
i@00000 CPU Device: Intel(R) Xeon(R) CPU E5-2698 v4 @ 2.20GHz
	Parsec Streams     : 20
	clockRate (GHz)    : 2.20
	peak Gflops        : double 176.0000, single 352.0000
i@00001 CPU Device: Intel(R) Xeon(R) CPU E5-2698 v4 @ 2.20GHz
	Parsec Streams     : 20
	clockRate (GHz)    : 2.20
	peak Gflops        : double 176.0000, single 352.0000
i@00000 Virtual Process Map with 1 VPs...
i@00000    Virtual Process of index 0 has 20 threads and cpuset 0x000fffff
i@00000     Thread 0 of VP 0 can be bound on cores 0x00000001
i@00000     Thread 1 of VP 0 can be bound on cores 0x00000002
i@00000     Thread 2 of VP 0 can be bound on cores 0x00000004
i@00000     Thread 3 of VP 0 can be bound on cores 0x00000008
i@00000     Thread 4 of VP 0 can be bound on cores 0x00000010
i@00000     Thread 5 of VP 0 can be bound on cores 0x00000020
i@00000     Thread 6 of VP 0 can be bound on cores 0x00000040
i@00000     Thread 7 of VP 0 can be bound on cores 0x00000080
i@00000     Thread 8 of VP 0 can be bound on cores 0x00000100
i@00000     Thread 9 of VP 0 can be bound on cores 0x00000200
i@00000     Thread 10 of VP 0 can be bound on cores 0x00000400
i@00000     Thread 11 of VP 0 can be bound on cores 0x00000800
i@00000     Thread 12 of VP 0 can be bound on cores 0x00001000
i@00000     Thread 13 of VP 0 can be bound on cores 0x00002000
i@00000     Thread 14 of VP 0 can be bound on cores 0x00004000
i@00000     Thread 15 of VP 0 can be bound on cores 0x00008000
i@00000     Thread 16 of VP 0 can be bound on cores 0x00010000
i@00000     Thread 17 of VP 0 can be bound on cores 0x00020000
i@00000     Thread 18 of VP 0 can be bound on cores 0x00040000
i@00000     Thread 19 of VP 0 can be bound on cores 0x00080000
i@00001 Virtual Process Map with 1 VPs...
i@00001    Virtual Process of index 0 has 20 threads and cpuset 0x000fffff
i@00001     Thread 0 of VP 0 can be bound on cores 0x00000001
i@00001     Thread 1 of VP 0 can be bound on cores 0x00000002
i@00001     Thread 2 of VP 0 can be bound on cores 0x00000004
i@00001     Thread 3 of VP 0 can be bound on cores 0x00000008
i@00001     Thread 4 of VP 0 can be bound on cores 0x00000010
i@00001     Thread 5 of VP 0 can be bound on cores 0x00000020
i@00001     Thread 6 of VP 0 can be bound on cores 0x00000040
i@00001     Thread 7 of VP 0 can be bound on cores 0x00000080
i@00001     Thread 8 of VP 0 can be bound on cores 0x00000100
i@00001     Thread 9 of VP 0 can be bound on cores 0x00000200
i@00001     Thread 10 of VP 0 can be bound on cores 0x00000400
i@00001     Thread 11 of VP 0 can be bound on cores 0x00000800
i@00001     Thread 12 of VP 0 can be bound on cores 0x00001000
i@00001     Thread 13 of VP 0 can be bound on cores 0x00002000
i@00001     Thread 14 of VP 0 can be bound on cores 0x00004000
i@00001     Thread 15 of VP 0 can be bound on cores 0x00008000
i@00001     Thread 16 of VP 0 can be bound on cores 0x00010000
i@00001     Thread 17 of VP 0 can be bound on cores 0x00020000
i@00001     Thread 18 of VP 0 can be bound on cores 0x00040000
i@00001     Thread 19 of VP 0 can be bound on cores 0x00080000
```

It looks like the two ranks are sharing the same cores from the output? 

A run of htop in a parallel terminal shows that only cores 0, 1, 13, 20, 21, 40, 41 and 65 have work to do (plus a bit for 77 and 46 sometimes), but definitely not all cores are active, and it's pretty slow.


Also, I tried to rebase the PR on the current master, but there are conflicts I'm not sure how to solve.

bosilca commented:
I was not able to find a way to translate between relative and absolute core numbering, so the reported bindings are relative to the allowed procs, and not absolute (as one would expect). Let me look again at the documentation to see if there is a way.

evaleev commented:
@bosilca ping ... need it badly :)

evaleev commented:
@bosilca ping again .. real showstopper

abouteiller commented on parsec/vpmap.c:
```diff
@@ -16,233 +16,264 @@
+static int parsec_nb_total_threads = 0;
 static int parse_binding_parameter(int vp, int nbth, char * binding);
 
-int vpmap_get_nb_total_threads(void)
```
This is used by dplasma, and is removed without replacement. The new `parsec_context_query(parsec_context_t*)` could be used instead but we don't have the parsec context at the caller site. I substituted with `parsec_vpmap_get_vp_thread(0)` in dplasma as that was probably the intent anyway (assuming vps have symmetrical number of threads). Should we reintroduce the function?

bosilca replied:
I brought it back.

abouteiller commented on parsec/parsec.c:
```diff
@@ -2584,10 +2560,11 @@ int parsec_check_overlapping_binding(parsec_context_t *context)
         int core = atoi(option);
-        if( (core > -1) && (core < parsec_hwloc_nb_real_cores()) )
-            context->comm_th_core = core;
+        /* negative core allowed to force an absolute core selection */
```
This is not true, passing a negative value to `find_core_by_idx` will cause an infinite loop.

abouteiller commented on parsec/parsec_hwloc.c:
```diff
@@ -78,43 +89,94 @@ static void parsec_hwloc_print_cpuset(int verb, char* msg, hwloc_cpuset_t cpuset
 {
 #if defined(PARSEC_HAVE_HWLOC)
-    if ( first_init ) {
+    if ( !parsec_hwloc_first_init ) {
```
This not being thread-safe/mutexed crashes the class/atomics ctest (only test were we exercise multithreaded initialization of hwloc with a straight call to bind threads without prior initialization in parsec_init)

abouteiller replied:
Discussion about why the simpler resolution is correct (extracted from Slack)
> Aurelien Bouteiller [17:31] we have this global `parsec_hwloc_first_init` but we do all sort of non atomic updates to it so it is useless
[17:34](https://icldisco.slack.com/archives/GC2L4J95G/p1709332445810949)
I can fix the class/atomics test (just call parsec_hwloc_init from the main thread before doing pthread_create) but I don’t see how this variable is supposed to help, if we have threads its broken, if we don’t its useless?
> 
>George Bosilca
  [11:16](https://icldisco.slack.com/archives/GC2L4J95G/p1709482567324649)
I think you misunderstood the goal of `parsec_hwloc_first_init`. In a normal situation, before threads can do anything related to HWLOC, parsec_init has been called, and the entire HWLOC subsystem has been initialized (so the parsec_hwloc_first_initis > 0)
[11:17](https://icldisco.slack.com/archives/GC2L4J95G/p1709482632527469)
parsec_hwloc_first_init protects against multiple initialization of the HWLOC info in case we create multiple contexts in the same application. There was no need to use atomics, because from where this is reached there is no race condition possible, in a normal setup

abouteiller commented:
The following command produces the correct binding for testing (as witnessed from hwloc-ls, the binding is correctly restricted by mpiexec to the correct sockets). '

```
PARSEC_MCA_device_cuda_memory_use=10 salloc -N1 -whexane  "/apps/spacks/2023-08-14/opt/spack/linux-rocky9-x86_64/gcc-11.3.1/openmpi-4.1.5-2rgaqk2wseegpmbdbbygvwrljccjaqsk/bin/mpiexec" --tag-output --oversubscribe -n 2 --bind-to socket --map-by socket --report-bindings hwloc-ls --restrict binding -c --no-io 
salloc: Granted job allocation 5788
[1,0]<stderr>:[hexane:3980861] MCW rank 0 bound to socket 0[core 0[hwt 0-1]], socket 0[core 1[hwt 0-1]], socket 0[core 2[hwt 0-1]], socket 0[core 3[hwt 0-1]], socket 0[core 4[hwt 0-1]], socket 0[core 5[hwt 0-1]], socket 0[core 6[hwt 0-1]], socket 0[core 7[hwt 0-1]]: [BB/BB/BB/BB/BB/BB/BB/BB][../../../../../../../..]
[1,1]<stderr>:[hexane:3980861] MCW rank 1 bound to socket 1[core 8[hwt 0-1]], socket 1[core 9[hwt 0-1]], socket 1[core 10[hwt 0-1]], socket 1[core 11[hwt 0-1]], socket 1[core 12[hwt 0-1]], socket 1[core 13[hwt 0-1]], socket 1[core 14[hwt 0-1]], socket 1[core 15[hwt 0-1]]: [../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/BB]
[1,0]<stdout>:Machine (62GB total) cpuset=0x55555555
[1,0]<stdout>:  Package L#0 cpuset=0x55555555
[1,0]<stdout>:    NUMANode L#0 (P#0 31GB) cpuset=0x55555555
[1,0]<stdout>:    L3 L#0 (12MB) cpuset=0x55555555
[1,0]<stdout>:      L2 L#0 (1280KB) cpuset=0x00010001
[1,0]<stdout>:        L1d L#0 (48KB) cpuset=0x00010001
[1,0]<stdout>:          L1i L#0 (32KB) cpuset=0x00010001
[1,0]<stdout>:            Core L#0 cpuset=0x00010001
[1,0]<stdout>:              PU L#0 (P#0) cpuset=0x00000001
[1,0]<stdout>:              PU L#1 (P#16) cpuset=0x00010000
[1,0]<stdout>:      L2 L#1 (1280KB) cpuset=0x00040004
[1,0]<stdout>:        L1d L#1 (48KB) cpuset=0x00040004
[1,0]<stdout>:          L1i L#1 (32KB) cpuset=0x00040004
[1,0]<stdout>:            Core L#1 cpuset=0x00040004
[1,0]<stdout>:              PU L#2 (P#2) cpuset=0x00000004
[1,0]<stdout>:              PU L#3 (P#18) cpuset=0x00040000
[1,0]<stdout>:      L2 L#2 (1280KB) cpuset=0x00100010
[1,0]<stdout>:        L1d L#2 (48KB) cpuset=0x00100010
[1,0]<stdout>:          L1i L#2 (32KB) cpuset=0x00100010
[1,0]<stdout>:            Core L#2 cpuset=0x00100010
[1,0]<stdout>:              PU L#4 (P#4) cpuset=0x00000010
[1,0]<stdout>:              PU L#5 (P#20) cpuset=0x00100000
[1,0]<stdout>:      L2 L#3 (1280KB) cpuset=0x00400040
[1,0]<stdout>:        L1d L#3 (48KB) cpuset=0x00400040
[1,0]<stdout>:          L1i L#3 (32KB) cpuset=0x00400040
[1,0]<stdout>:            Core L#3 cpuset=0x00400040
[1,0]<stdout>:              PU L#6 (P#6) cpuset=0x00000040
[1,0]<stdout>:              PU L#7 (P#22) cpuset=0x00400000
[1,0]<stdout>:      L2 L#4 (1280KB) cpuset=0x01000100
[1,0]<stdout>:        L1d L#4 (48KB) cpuset=0x01000100
[1,0]<stdout>:          L1i L#4 (32KB) cpuset=0x01000100
[1,0]<stdout>:            Core L#4 cpuset=0x01000100
[1,0]<stdout>:              PU L#8 (P#8) cpuset=0x00000100
[1,0]<stdout>:              PU L#9 (P#24) cpuset=0x01000000
[1,0]<stdout>:      L2 L#5 (1280KB) cpuset=0x04000400
[1,0]<stdout>:        L1d L#5 (48KB) cpuset=0x04000400
[1,0]<stdout>:          L1i L#5 (32KB) cpuset=0x04000400
[1,0]<stdout>:            Core L#5 cpuset=0x04000400
[1,0]<stdout>:              PU L#10 (P#10) cpuset=0x00000400
[1,0]<stdout>:              PU L#11 (P#26) cpuset=0x04000000
[1,0]<stdout>:      L2 L#6 (1280KB) cpuset=0x10001000
[1,0]<stdout>:        L1d L#6 (48KB) cpuset=0x10001000
[1,0]<stdout>:          L1i L#6 (32KB) cpuset=0x10001000
[1,0]<stdout>:            Core L#6 cpuset=0x10001000
[1,0]<stdout>:              PU L#12 (P#12) cpuset=0x00001000
[1,0]<stdout>:              PU L#13 (P#28) cpuset=0x10000000
[1,0]<stdout>:      L2 L#7 (1280KB) cpuset=0x40004000
[1,0]<stdout>:        L1d L#7 (48KB) cpuset=0x40004000
[1,0]<stdout>:          L1i L#7 (32KB) cpuset=0x40004000
[1,0]<stdout>:            Core L#7 cpuset=0x40004000
[1,0]<stdout>:              PU L#14 (P#14) cpuset=0x00004000
[1,0]<stdout>:              PU L#15 (P#30) cpuset=0x40000000
[1,0]<stdout>:  Package L#1 cpuset=0x0
[1,0]<stdout>:    NUMANode L#1 (P#1 31GB) cpuset=0x0
[1,1]<stdout>:Machine (62GB total) cpuset=0xaaaaaaaa
[1,1]<stdout>:  Package L#0 cpuset=0xaaaaaaaa
[1,1]<stdout>:    NUMANode L#0 (P#1 31GB) cpuset=0xaaaaaaaa
[1,1]<stdout>:    L3 L#0 (12MB) cpuset=0xaaaaaaaa
[1,1]<stdout>:      L2 L#0 (1280KB) cpuset=0x00020002
[1,1]<stdout>:        L1d L#0 (48KB) cpuset=0x00020002
[1,1]<stdout>:          L1i L#0 (32KB) cpuset=0x00020002
[1,1]<stdout>:            Core L#0 cpuset=0x00020002
[1,1]<stdout>:              PU L#0 (P#1) cpuset=0x00000002
[1,1]<stdout>:              PU L#1 (P#17) cpuset=0x00020000
[1,1]<stdout>:      L2 L#1 (1280KB) cpuset=0x00080008
[1,1]<stdout>:        L1d L#1 (48KB) cpuset=0x00080008
[1,1]<stdout>:          L1i L#1 (32KB) cpuset=0x00080008
[1,1]<stdout>:            Core L#1 cpuset=0x00080008
[1,1]<stdout>:              PU L#2 (P#3) cpuset=0x00000008
[1,1]<stdout>:              PU L#3 (P#19) cpuset=0x00080000
[1,1]<stdout>:      L2 L#2 (1280KB) cpuset=0x00200020
[1,1]<stdout>:        L1d L#2 (48KB) cpuset=0x00200020
[1,1]<stdout>:          L1i L#2 (32KB) cpuset=0x00200020
[1,1]<stdout>:            Core L#2 cpuset=0x00200020
[1,1]<stdout>:              PU L#4 (P#5) cpuset=0x00000020
[1,1]<stdout>:              PU L#5 (P#21) cpuset=0x00200000
[1,1]<stdout>:      L2 L#3 (1280KB) cpuset=0x00800080
[1,1]<stdout>:        L1d L#3 (48KB) cpuset=0x00800080
[1,1]<stdout>:          L1i L#3 (32KB) cpuset=0x00800080
[1,1]<stdout>:            Core L#3 cpuset=0x00800080
[1,1]<stdout>:              PU L#6 (P#7) cpuset=0x00000080
[1,1]<stdout>:              PU L#7 (P#23) cpuset=0x00800000
[1,1]<stdout>:      L2 L#4 (1280KB) cpuset=0x02000200
[1,1]<stdout>:        L1d L#4 (48KB) cpuset=0x02000200
[1,1]<stdout>:          L1i L#4 (32KB) cpuset=0x02000200
[1,1]<stdout>:            Core L#4 cpuset=0x02000200
[1,1]<stdout>:              PU L#8 (P#9) cpuset=0x00000200
[1,1]<stdout>:              PU L#9 (P#25) cpuset=0x02000000
[1,1]<stdout>:      L2 L#5 (1280KB) cpuset=0x08000800
[1,1]<stdout>:        L1d L#5 (48KB) cpuset=0x08000800
[1,1]<stdout>:          L1i L#5 (32KB) cpuset=0x08000800
[1,1]<stdout>:            Core L#5 cpuset=0x08000800
[1,1]<stdout>:              PU L#10 (P#11) cpuset=0x00000800
[1,1]<stdout>:              PU L#11 (P#27) cpuset=0x08000000
[1,1]<stdout>:      L2 L#6 (1280KB) cpuset=0x20002000
[1,1]<stdout>:        L1d L#6 (48KB) cpuset=0x20002000
[1,1]<stdout>:          L1i L#6 (32KB) cpuset=0x20002000
[1,1]<stdout>:            Core L#6 cpuset=0x20002000
[1,1]<stdout>:              PU L#12 (P#13) cpuset=0x00002000
[1,1]<stdout>:              PU L#13 (P#29) cpuset=0x20000000
[1,1]<stdout>:      L2 L#7 (1280KB) cpuset=0x80008000
[1,1]<stdout>:        L1d L#7 (48KB) cpuset=0x80008000
[1,1]<stdout>:          L1i L#7 (32KB) cpuset=0x80008000
[1,1]<stdout>:            Core L#7 cpuset=0x80008000
[1,1]<stdout>:              PU L#14 (P#15) cpuset=0x00008000
[1,1]<stdout>:              PU L#15 (P#31) cpuset=0x80000000
[1,1]<stdout>:  Package L#1 cpuset=0x0
[1,1]<stdout>:    NUMANode L#1 (P#0 31GB) cpuset=0x0
salloc: Relinquishing job allocation 5788
```

The resultant binding in Parsec does not appear correct: 

```
PARSEC_MCA_device_cuda_memory_use=10 salloc -N1 -whexane  "/apps/spacks/2023-08-14/opt/spack/linux-rocky9-x86_64/gcc-11.3.1/openmpi-4.1.5-2rgaqk2wseegpmbdbbygvwrljccjaqsk/bin/mpiexec" --tag-output --oversubscribe -n 2 --bind-to socket --map-by socket --report-bindings build.cuda/parsec/tests/api/init_fini --mca runtime_report_bindings 1
salloc: Granted job allocation 5790
[1,0]<stderr>:[hexane:3981005] MCW rank 0 bound to socket 0[core 0[hwt 0-1]], socket 0[core 1[hwt 0-1]], socket 0[core 2[hwt 0-1]], socket 0[core 3[hwt 0-1]], socket 0[core 4[hwt 0-1]], socket 0[core 5[hwt 0-1]], socket 0[core 6[hwt 0-1]], socket 0[core 7[hwt 0-1]]: [BB/BB/BB/BB/BB/BB/BB/BB][../../../../../../../..]
[1,1]<stderr>:[hexane:3981005] MCW rank 1 bound to socket 1[core 8[hwt 0-1]], socket 1[core 9[hwt 0-1]], socket 1[core 10[hwt 0-1]], socket 1[core 11[hwt 0-1]], socket 1[core 12[hwt 0-1]], socket 1[core 13[hwt 0-1]], socket 1[core 14[hwt 0-1]], socket 1[core 15[hwt 0-1]]: [../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/BB]
[1,1]<stdout>:Process binding [rank 0]: cpuset [ALLOWED  ]: 0xaaaaaaaa
[1,1]<stdout>:Process binding [rank 0]: cpuset [USED     ]: 0x0000aaaa
[1,1]<stdout>:Process binding [rank 0]: cpuset [FREE     ]: 0xaaaa0000
[1,0]<stdout>:Process binding [rank 0]: cpuset [ALLOWED  ]: 0x55555555
[1,0]<stdout>:Process binding [rank 0]: cpuset [USED     ]: 0x00005555
[1,0]<stdout>:Process binding [rank 0]: cpuset [FREE     ]: 0x55550000
[1,0]<stderr>:W@00000 /!\ DEBUG LEVEL WILL PROBABLY REDUCE THE PERFORMANCE OF THIS RUN /!\.
[1,0]<stderr>:i@00000 Virtual Process Map with 1 VPs...
[1,0]<stderr>:i@00000    Virtual Process of index 0 has 8 threads and logical cpuset 0x000000ff
[1,0]<stderr>:           physical cpuset 0x55555555
[1,0]<stderr>:i@00000     Thread 0 of VP 0 can be bound on logical cores 0x00000001 (physical cores 0x00010001)
[1,0]<stderr>:i@00000     Thread 1 of VP 0 can be bound on logical cores 0x00000002 (physical cores 0x00040004)
[1,0]<stderr>:i@00000     Thread 2 of VP 0 can be bound on logical cores 0x00000004 (physical cores 0x00100010)
[1,0]<stderr>:i@00000     Thread 3 of VP 0 can be bound on logical cores 0x00000008 (physical cores 0x00400040)
[1,0]<stderr>:i@00000     Thread 4 of VP 0 can be bound on logical cores 0x00000010 (physical cores 0x01000100)
[1,0]<stderr>:i@00000     Thread 5 of VP 0 can be bound on logical cores 0x00000020 (physical cores 0x04000400)
[1,0]<stderr>:i@00000     Thread 6 of VP 0 can be bound on logical cores 0x00000040 (physical cores 0x10001000)
[1,0]<stderr>:i@00000     Thread 7 of VP 0 can be bound on logical cores 0x00000080 (physical cores 0x40004000)
[1,1]<stderr>:i@00001 Virtual Process Map with 1 VPs...
[1,1]<stderr>:i@00001    Virtual Process of index 0 has 8 threads and logical cpuset 0x000000ff
[1,1]<stderr>:           physical cpuset 0x55555555
[1,1]<stderr>:i@00001     Thread 0 of VP 0 can be bound on logical cores 0x00000001 (physical cores 0x00010001)
[1,1]<stderr>:i@00001     Thread 1 of VP 0 can be bound on logical cores 0x00000002 (physical cores 0x00040004)
[1,1]<stderr>:i@00001     Thread 2 of VP 0 can be bound on logical cores 0x00000004 (physical cores 0x00100010)
[1,1]<stderr>:i@00001     Thread 3 of VP 0 can be bound on logical cores 0x00000008 (physical cores 0x00400040)
[1,1]<stderr>:i@00001     Thread 4 of VP 0 can be bound on logical cores 0x00000010 (physical cores 0x01000100)
[1,1]<stderr>:i@00001     Thread 5 of VP 0 can be bound on logical cores 0x00000020 (physical cores 0x04000400)
[1,1]<stderr>:i@00001     Thread 6 of VP 0 can be bound on logical cores 0x00000040 (physical cores 0x10001000)
[1,1]<stderr>:i@00001     Thread 7 of VP 0 can be bound on logical cores 0x00000080 (physical cores 0x40004000)
salloc: Relinquishing job allocation 5790
```

abouteiller commented:
It looks like the message above is partially misleading: we initialized the vpmap before we extract the ALLOWED mask, so the vpmap_from_flat initializes something that is not the same as what the real final binding is. This is a problem because using VPs will do something different than not using VPs, and the output is misleading, but the actual binding produced should be correct.

abouteiller commented:
Looks like inherited binding is correct (beside the problem with the vpmap above, comparing `--bind-to none -c 8` vs `--bind-to socket` results in double the performance and using 16 hw cores in the second case and not having the message about running oversubscribed. 

```
 PARSEC_MCA_device_cuda_memory_use=10 salloc -N1 --ntasks-per-node=2 -whexane  "/apps/spacks/2023-08-14/opt/spack/linux-rocky9-x86_64/gcc-11.3.1/openmpi-4.1.5-2rgaqk2wseegpmbdbbygvwrljccjaqsk/bin/mpiexec" --tag-output -n 2 --bind-to none --map-by socket --report-bindings build.cuda/tests/testing_dpotrf -g 0 -N 30000 -x -v=4 -t 384 -c 8

[1,1]<stderr>:W@00001 /!\ PERFORMANCE MIGHT BE REDUCED /!\: Multiple PaRSEC processes on the same node may share the same physical core(s);
[1,1]<stderr>:  This is often unintentional, and will perform poorly.
[1,1]<stderr>:  Note that in managed environments (e.g., ALPS, jsrun), the launcher may set `cgroups`
[1,1]<stderr>:  and hide the real binding from PaRSEC; if you verified that the binding is correct,
[1,1]<stderr>:  this message can be silenced using the MCA argument `runtime_warn_slow_binding`.
[1,0]<stderr>:#+++++ cores detected       : 8
[1,0]<stderr>:#+++++ nodes x cores + gpu  : 2 x 8 + 0 (16+0)
[1,0]<stdout>:[****] TIME(s)     31.77802 : dpotrf      PxQxg=   2 1   0 NB=  384 N=   30000 :     283.228825 gflops - ENQ&PROG&DEST     31.80433 :     282.994460 gflops - ENQ      0.02591 - DEST      0.00040

...

 PARSEC_MCA_device_cuda_memory_use=10 salloc -N1 --ntasks-per-node=2 -whexane  "/apps/spacks/2023-08-14/opt/spack/linux-rocky9-x86_64/gcc-11.3.1/openmpi-4.1.5-2rgaqk2wseegpmbdbbygvwrljccjaqsk/bin/mpiexec" --tag-output -n 2 --bind-to socket --map-by socket --report-bindings build.cuda/tests/testing_dpotrf -g 0 -N 30000 -x -v=4 -t 384
[1,0]<stderr>:#+++++ cores detected       : 8
[1,0]<stderr>:#+++++ nodes x cores + gpu  : 2 x 8 + 0 (16+0)
[1,0]<stdout>:[****] TIME(s)     19.26431 : dpotrf      PxQxg=   2 1   0 NB=  384 N=   30000 :     467.208496 gflops - ENQ&PROG&DEST     19.26601 :     467.167396 gflops - ENQ      0.00008 - DEST      0.00161
```

abouteiller commented:
vpmap initialization creates and fills `                parsec_vpmap[vp].threads[t+ht].cpuset = HWLOC_ALLOC();`  with all sorts of intricate things (that are not abiding with the restricted mask) but these are write only variables.

abouteiller committed:
Disable the vpmap display of binding parameters, and warn that these are
write only atm. Use the bindthread to print the actual binding effected, if
requested.

bosilca committed:
Add documentation about the thread bindings.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

abouteiller commented on parsec/parsec_hwloc.c:
```diff
@@ -78,43 +89,94 @@ static void parsec_hwloc_print_cpuset(int verb, char* msg, hwloc_cpuset_t cpuset
+
+    if( 0 == parsec_available_binding_resources ) {  /* bozo case: something wrong let's just assume single core */
+        parsec_warning("Could not identify the CPU sets to be used. Fall back on safety more, single core");
+        parsec_available_binding_resources = 1;
```
Should we fallback to no-binding if we don't know what to do?

bosilca replied:
something bad here means we are unable to get a clear number of cores from hwloc. I doubt we will be able to bind to anything, so assuming single core sounds like a valid approach. In any case, this is extremely unlikely to happen.

abouteiller commented on parsec/parsec.c:
```diff
@@ -468,42 +485,53 @@ parsec_context_t* parsec_init( int nb_cores, int* pargc, char** pargv[] )
+    parsec_mca_param_reg_int_name("runtime", "report_bindings",
+                                  "Report the binding of all PaRSEC resources (main thread, worker threads, and communication thread).",
+                                  false, false, parsec_report_bindings, &parsec_report_bindings);
+    parsec_mca_param_reg_int_name("runtime", "ignore_bindings",
```
```suggestion
    parsec_mca_param_reg_int_name("runtime", "ignore_external_bindings",
```

abouteiller commented on parsec/vpmap.c:
```diff
@@ -16,233 +16,281 @@
+        parsec_vpmap[i].cpuset = HWLOC_ALLOC();
+        for( int j = 0; j < parsec_vpmap[i].nbthreads; j++ ) {
+            if( NULL == parsec_vpmap[i].threads[j].cpuset ) {
+                parsec_vpmap[i].threads[j].cpuset = HWLOC_ALLOC();
```
@bosilca Can you confirm that these are not write only variables anymore, the confusion with these being set but never used is the reason the PR was not merged before.

bosilca replied:
They are now the one driving the binding.
